{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "athletic-petersburg",
   "metadata": {},
   "source": [
    "# Documentation - piRNA pipe\n",
    "\n",
    "\n",
    "\n",
    "## Report\n",
    "\n",
    "### 1. Expression \n",
    "\n",
    "1.1 Distribution of small RNAs, (Julius et al., 2008, Science, Figure2A)    \n",
    "1.2 plots, read coverage on TE  \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## 1 Purpose  \n",
    "\n",
    "For small RNA seqeuencing analysis (total, piwi-IP seq, etc). \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## 2 workflow  \n",
    "\n",
    "\n",
    "### 2.1 pre-analysis  \n",
    "\n",
    "+ Main steps \n",
    "\n",
    "  - remove structural RNAs (rRNA, tRNA, snRNA, snoRNA, ..., check rRNA%) \n",
    "  - remove miRNAs (miRNA hairpin, 0 mismatch, unique + multi)  \n",
    "  - split by length ([23,29], others)  \n",
    "  - map to TE consensus (0 mismatch, unique) \n",
    "  - map to piRNA clusters (0 mismatch, unique)  \n",
    "  - map to genome (0 mismatch, unique + multi)  \n",
    "  \n",
    "+ Extra  \n",
    "\n",
    "  - read size distribution  \n",
    "  - evaluate 1U-10A   \n",
    "  - check overlap between samples  \n",
    "  \n",
    "+ To-do  \n",
    "  \n",
    "  - extract siRNAs (21nt)  \n",
    "  - ping-pong signature score \n",
    "  - correlation between samples\n",
    "  - compare with piRBase (v2.0), other published data, overlap (+)\n",
    "  \n",
    "  - TE read count\n",
    " \n",
    "\n",
    "\n",
    "\n",
    "## 3 Analysis  \n",
    "\n",
    "\n",
    "### 3.1 main  \n",
    "\n",
    "  - `piRNA_pipe.py`  \n",
    "  \n",
    "  \n",
    "  \n",
    "### 3.2 modules  \n",
    "\n",
    "  - `fragsize.py` - calculate the distribution of read length  \n",
    "  \n",
    "  - `u1a10.py` - calculate the U1A10 content  \n",
    "  \n",
    "  - `fx_overlap.py` - check the overlap between two fastx file\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### 3.3 Directory structure \n",
    "\n",
    "```\n",
    "├── 00.raw_data\n",
    "├── 01.clean_data\n",
    "├── 02.collapse\n",
    "├── 03.overlap\n",
    "├── 04.smRNA\n",
    "├── 05.miRNA\n",
    "├── 06.size_exclude\n",
    "├── 06.size_select\n",
    "├── 07.te\n",
    "├── 08.piRNA_cluster\n",
    "├── 09.genome\n",
    "├── 10.unmap\n",
    "├── 11.stat\n",
    "└── 12.report\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "### 3.4 Output files\n",
    "\n",
    "\n",
    "\n",
    " \n",
    "## x. Change log\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "```\n",
    "piRNA analysis (small RNAseq) \n",
    "\n",
    "\n",
    "date: 2020-12-27\n",
    "\n",
    "1. support collapsed reads, counting, RNA seqs, reads\n",
    "\n",
    "\n",
    "date: 2020-12-23\n",
    "Author: Ming Wang\n",
    "\n",
    "# flowchart of piRNA analysis\n",
    "1. remove structural RNAs (uniqe + multi)\n",
    "2. remove miRNAs (unique + multi) \n",
    "3. remove reads not in [23-29] nt \n",
    "4. collapse reads: consider 1-23nt only, allow 1-2 at 3' differ (2019-11-26)\n",
    "5. split into 4 groups: (1U+/-,10A+/-;)\n",
    "6. map to TE consensus, (unique + multi), only 1U_not_10A\n",
    "7. map to genome (unique + multi) \n",
    "Functions:\n",
    "rename fastq reads: piR0000001-0000001\n",
    "piR (piRNA), (piRNA number) {reads number}\n",
    "mapping, unique + multiple/unique\n",
    "\n",
    "\n",
    "version: 2020-07-28\n",
    "update:\n",
    "1. remove temp fastq files\n",
    "2. gzip fastq files\n",
    "\n",
    "version: 2020-07-25\n",
    "update:\n",
    "1. collapse reads \n",
    "\n",
    "\n",
    "date: 2020-07-23\n",
    "in brief:\n",
    "\n",
    "1. remove 3' adapters  \n",
    "2. remove structural RNAs (tRNAs, rRNAs) (unique + multiple)     \n",
    "3. filt by length, 23-29 nt   \n",
    "4. collapse reads, only compare 1-23 nt (*)  \n",
    "   collapse reads, (regular)  \n",
    "   trim reads to 23nt from 3' end  \n",
    "\n",
    "5. map to TE consensus (unique, multiple, unique + multiple)   \n",
    "6. map to piRNA clusters (unique, multiple, unique + multiple)   \n",
    "7. map to genome (not-TE, no-piRNAcluster)    \n",
    "8. Overall, map to genome\n",
    "```\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "brutal-detroit",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "from piRNA_pipe import *\n",
    "\n",
    "args = {\n",
    "    'fq': 'test/a.fq.gz',\n",
    "    'outdir': 'test/results',\n",
    "    'collapse': True,\n",
    "    'subject': 'test/b.fq.gz',\n",
    "    'force_overlap': True,\n",
    "    'ov_type': 1,\n",
    "}\n",
    "# p = pipe(**args).run()\n",
    "\n",
    "args = {\n",
    "    'query': 'test/a.fq',\n",
    "    'subject': 'test/b.fq',\n",
    "    'outdir': 'test/overlap',\n",
    "    'mm': 2\n",
    "}\n",
    "OverlapFq(**args).run()\n",
    "# overlap_fq('test/a.fq', 'test/b.fq', 'test/overlap2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "protecting-incentive",
   "metadata": {},
   "outputs": [],
   "source": [
    "# U1A10\n",
    "\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "from fastx import *\n",
    "\n",
    "fx = ['test/a.fq.gz', 'test/b.fq.gz']\n",
    "\n",
    "FxU1A10(fx).run()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "instant-prairie",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alignemt\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import pyfastx\n",
    "from align import *\n",
    "\n",
    "args = {\n",
    "    'fx': 'test/a.fq.gz',\n",
    "    'index': '/data/yulab/wangming/data/genome/dm6/bowtie_index/miRNA',\n",
    "    'outdir': 'test/bbbbbb',\n",
    "    'unique': 'multi'\n",
    "}\n",
    "\n",
    "Align(**args).run()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "offshore-header",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PiRNApipe\n",
    "\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import pyfastx\n",
    "from piRNA_pipe import *\n",
    "\n",
    "args = {\n",
    "    'fq': 'test/dme.1m.fa.gz',\n",
    "    'outdir': 'test/aaaaaa',\n",
    "    'genome': 'dm6',\n",
    "    'subject_list': 'test/a.fq.gz'\n",
    "}\n",
    "\n",
    "p1 = PiRNApipe(**args)\n",
    "# p1.run()\n",
    "# p2 = overlap_subject(p1)\n",
    "\n",
    "for k, v in p1.__dict__.items():\n",
    "    print('{}: {}'.format(k, v))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "creative-genius",
   "metadata": {},
   "outputs": [],
   "source": [
    "# stat\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "from qc import *\n",
    "import qc\n",
    "from hiseq.fragsize.fragsize import BamFragSize\n",
    "\n",
    "# x = 'test/aaaaaa/a'\n",
    "# y = x + '/00.raw_data'\n",
    "# df = PiRNApipeStat(x).run()\n",
    "# df\n",
    "\n",
    "# x = '/data/yulab/wangming/work/yu_2020/public_data/2016_molcell_siomi/results/smRNA/smRNAseq_H1_KD_Piwi_IP_rep1'\n",
    "\n",
    "# p = PiRNApipeStat(x)\n",
    "# # p.run()\n",
    "# # p.report()\n",
    "# bam_list = p.list_bam(u1a10=True)\n",
    "# # bam_list\n",
    "# for b in bam_list:\n",
    "#     try:\n",
    "#         FxFragSize(b).run()\n",
    "#     except:\n",
    "#         print(\"!AAAA-1\", b)\n",
    "\n",
    "x = '/data/yulab/wangming/work/yu_2020/public_data/2008_science_greg/results/piRNA_pipe/Embryo_0_2hrs_rep1_w1118_0_2hr_Emb'\n",
    "p = PiRNApipeStat(x)\n",
    "# # p.run()\n",
    "# dirs = p.list_dir(u1a10=False, group=False)\n",
    "# for d in dirs:\n",
    "#     try: \n",
    "#         p.stat_dir(d)\n",
    "#     except:\n",
    "#         print('!AAAA-1', d)\n",
    "d = '/data/yulab/wangming/work/yu_2020/public_data/2008_science_greg/results/piRNA_pipe/Embryo_0_2hrs_rep1_w1118_0_2hr_Emb/00.raw_data'\n",
    "p.stat_dir(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "determined-format",
   "metadata": {},
   "outputs": [],
   "source": [
    "def stat_dir(d):\n",
    "    \"\"\"Create summary for level-1 directory\n",
    "    number of reads, for each category\n",
    "    total: \n",
    "    U1_A10: \n",
    "    \"\"\"\n",
    "    h = ['sample', 'num_reads', 'num_seqs']\n",
    "    dfx = pd.DataFrame(columns=h)\n",
    "    # level-1: dirs\n",
    "    group = os.path.basename(d)\n",
    "    t1 = self.list_count_toml(d)\n",
    "    if len(t1) == 1:\n",
    "        df1 = self.load_toml(t1[0])\n",
    "    else:\n",
    "        df1 = dfx\n",
    "    df1.columns = h\n",
    "    df1['u1a10'] = 'all'\n",
    "    # level-2: U1A10\n",
    "    t2 = self.list_count_toml(d+'/U1_A10')\n",
    "    frames = [self.load_toml(i) for i in t2]\n",
    "    if len(frames) > 0:\n",
    "        df2 = pd.concat(frames, axis=0)\n",
    "        df2.columns = h\n",
    "        df2[['sample', 'u1a10']] = df2['sample'].str.split('.', 1, expand=True)\n",
    "    else:\n",
    "        df2 = dfx\n",
    "        df2['u1a10'] = 'all'\n",
    "    # combine\n",
    "    df = pd.concat([df1, df2], 'index')\n",
    "    df['group'] = group\n",
    "    # stat\n",
    "    if len(t1) > 0:\n",
    "        d_csv = os.path.join(d, 'fx_stat.csv')\n",
    "        df.to_csv(d_csv, index=False)\n",
    "    return df\n",
    "\n",
    "d = '/data/yulab/wangming/work/yu_2020/public_data/2008_science_greg/results/piRNA_pipe/Embryo_0_2hrs_rep1_w1118_0_2hr_Emb/07.te/uniuq/U1_A10'\n",
    "t2 = p.list_count_toml(d+'/U1_A10')\n",
    "frames = [p.load_toml(i) for i in t2]\n",
    "# if len(frames) > 0:\n",
    "#     df2 = pd.concat(frames, axis=0)\n",
    "#     df2.columns = h\n",
    "#     df2[['sample', 'u1a10']] = df2['sample'].str.split('.', 1, expand=True)\n",
    "# else:\n",
    "#     df2 = dfx\n",
    "#     df2['u1a10'] = 'all'\n",
    "# t1 = p.list_count_toml(d)\n",
    "# df1 = p.load_toml(t1[0])\n",
    "# df1\n",
    "pd.concat(frames, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "international-resort",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = ['abc.txt.gz', 'a.fq.gz', 'b.unmap.fq.gz']\n",
    "\n",
    "\n",
    "def get_fx_name(x):\n",
    "    \"\"\"Parse the name of fastx\n",
    "    a.fq\n",
    "    a.fq.gz\n",
    "    a.unmap.fq\n",
    "    a.unmap.fq.gz\n",
    "    \"\"\"\n",
    "    if isinstance(x, str):\n",
    "        x_name = os.path.basename(x)\n",
    "        # remove .gz\n",
    "        if x_name[-3:] == '.gz':\n",
    "            x_name = x_name[:-3]\n",
    "        # remove ext:\n",
    "        x_name = os.path.splitext(x_name)[0]\n",
    "        # remove unmap\n",
    "        if x_name[-6:] == '.unmap':\n",
    "            x_name = x_name[:-6]\n",
    "        return x_name\n",
    "    \n",
    "list(map(get_fx_name, a))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "compliant-conservative",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "prompt-nebraska",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-03-12T03:47:48.912374Z",
     "iopub.status.busy": "2021-03-12T03:47:48.911829Z",
     "iopub.status.idle": "2021-03-12T03:47:49.004924Z",
     "shell.execute_reply": "2021-03-12T03:47:49.004123Z",
     "shell.execute_reply.started": "2021-03-12T03:47:48.912322Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------------------------------\n",
      "track_file: /data/yulab/wangming/work/devel_pipeline/pirna_seq/scripts/test/aaaaaa/a/07.te/unique/a.bigwig,\n",
      "/data/yulab/wangming/work/devel_pipeline/pirna_seq/scripts/test/aaaaaa/b/07.te/unique/b.bigwig\n",
      "outdir: test/ccc\n",
      "region_list: /data/yulab/wangming/work/devel_pipeline/pirna_seq/scripts/test/aaaaaa/a/07.te/unique/a.bigwig\n",
      "--------------------------------------------------------------------------------\n",
      "1/126 - Idefix\n",
      "[2021-03-12 11:47:49 INFO] get_track() skipped, file exists: /data/yulab/wangming/work/devel_pipeline/pirna_seq/scripts/test/ccc/Idefix/Idefix_1_7411.png\n"
     ]
    }
   ],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "from piRNA_pipe import *\n",
    "from utils import *\n",
    "from track_file import *\n",
    "\n",
    "x = ['test/aaaaaa/a', 'test/aaaaaa/b']\n",
    "\n",
    "\n",
    "# run_bam_to_bw(x, 'te')\n",
    "# get_x_file(x, 'bigwig', check_exists=False)\n",
    "\n",
    "bw = [get_x_file(i, 'bigwig', 'te') for i in x]\n",
    "outdir = 'test/ccc'\n",
    "\n",
    "get_tracks(bw, outdir=outdir)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "constitutional-intent",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
