{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "large-uzbekistan",
   "metadata": {},
   "source": [
    "# Documentation - piRNA pipe\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## 1 Purpose  \n",
    "\n",
    "For small RNA seqeuencing analysis (total, piwi-IP seq, etc). \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## 2 workflow  \n",
    "\n",
    "\n",
    "### 2.1 pre-analysis  \n",
    "\n",
    "+ Main steps \n",
    "\n",
    "  - remove structural RNAs (rRNA, tRNA, snRNA, snoRNA, ..., check rRNA%) \n",
    "  - remove miRNAs (miRNA hairpin, 0 mismatch, unique + multi)  \n",
    "  - split by length ([23,29], others)  \n",
    "  - map to TE consensus (0 mismatch, unique) \n",
    "  - map to piRNA clusters (0 mismatch, unique)  \n",
    "  - map to genome (0 mismatch, unique + multi)  \n",
    "  \n",
    "+ Extra  \n",
    "\n",
    "  - read size distribution  \n",
    "  - evaluate 1U-10A   \n",
    "  - check overlap between samples  \n",
    "  \n",
    "+ To-do  \n",
    "  \n",
    "  - extract siRNAs (21nt)  \n",
    "  - ping-pong signature score \n",
    "  - correlation between samples\n",
    "  - compare with piRBase (v2.0), other published data\n",
    " \n",
    "\n",
    "\n",
    "\n",
    "## 3 Analysis  \n",
    "\n",
    "\n",
    "### 3.1 main  \n",
    "\n",
    "  - `piRNA_pipe.py`  \n",
    "  \n",
    "  \n",
    "  \n",
    "### 3.2 modules  \n",
    "\n",
    "  - `fragsize.py` - calculate the distribution of read length  \n",
    "  \n",
    "  - `u1a10.py` - calculate the U1A10 content  \n",
    "  \n",
    "  - `fx_overlap.py` - check the overlap between two fastx file\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### 3.3 Directory structure \n",
    "\n",
    "```\n",
    "├── 00.raw_data\n",
    "├── 01.clean_data\n",
    "├── 02.collapse\n",
    "├── 03.overlap\n",
    "├── 04.smRNA\n",
    "├── 05.miRNA\n",
    "├── 06.size_exclude\n",
    "├── 06.size_select\n",
    "├── 07.te\n",
    "├── 08.piRNA_cluster\n",
    "├── 09.genome\n",
    "├── 10.unmap\n",
    "├── 11.stat\n",
    "└── 12.report\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "### 3.4 Output files\n",
    "\n",
    "\n",
    "\n",
    " \n",
    "## x. Change log\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "```\n",
    "piRNA analysis (small RNAseq) \n",
    "\n",
    "\n",
    "date: 2020-12-27\n",
    "\n",
    "1. support collapsed reads, counting, RNA seqs, reads\n",
    "\n",
    "\n",
    "date: 2020-12-23\n",
    "Author: Ming Wang\n",
    "\n",
    "# flowchart of piRNA analysis\n",
    "1. remove structural RNAs (uniqe + multi)\n",
    "2. remove miRNAs (unique + multi) \n",
    "3. remove reads not in [23-29] nt \n",
    "4. collapse reads: consider 1-23nt only, allow 1-2 at 3' differ (2019-11-26)\n",
    "5. split into 4 groups: (1U+/-,10A+/-;)\n",
    "6. map to TE consensus, (unique + multi), only 1U_not_10A\n",
    "7. map to genome (unique + multi) \n",
    "Functions:\n",
    "rename fastq reads: piR0000001-0000001\n",
    "piR (piRNA), (piRNA number) {reads number}\n",
    "mapping, unique + multiple/unique\n",
    "\n",
    "\n",
    "version: 2020-07-28\n",
    "update:\n",
    "1. remove temp fastq files\n",
    "2. gzip fastq files\n",
    "\n",
    "version: 2020-07-25\n",
    "update:\n",
    "1. collapse reads \n",
    "\n",
    "\n",
    "date: 2020-07-23\n",
    "in brief:\n",
    "\n",
    "1. remove 3' adapters  \n",
    "2. remove structural RNAs (tRNAs, rRNAs) (unique + multiple)     \n",
    "3. filt by length, 23-29 nt   \n",
    "4. collapse reads, only compare 1-23 nt (*)  \n",
    "   collapse reads, (regular)  \n",
    "   trim reads to 23nt from 3' end  \n",
    "\n",
    "5. map to TE consensus (unique, multiple, unique + multiple)   \n",
    "6. map to piRNA clusters (unique, multiple, unique + multiple)   \n",
    "7. map to genome (not-TE, no-piRNAcluster)    \n",
    "8. Overall, map to genome\n",
    "```\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "entertaining-fault",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-03-03T07:57:19.850316Z",
     "iopub.status.busy": "2021-03-03T07:57:19.849790Z",
     "iopub.status.idle": "2021-03-03T07:57:20.658215Z",
     "shell.execute_reply": "2021-03-03T07:57:20.657599Z",
     "shell.execute_reply.started": "2021-03-03T07:57:19.850265Z"
    }
   },
   "outputs": [],
   "source": [
    "from piRNA_pipe import *\n",
    "\n",
    "\n",
    "args = {\n",
    "    ''\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
